paramThreeToGL made public

spite opened the pull request:
This change is useful -and necessary- to get data from three.js objects and operate with them with the WebGL API directly.

spite committed:
paramThreeToGL made public

mrdoob commented:
I assume this is for https://github.com/spite/THREE.UpdatableTexture? Was this the only change you have to do?

I've always seen `paramThreeToGL` as a hack so exposing doesn't feel good. Maybe that code could be modularised into something like `THREE.WebGLUtils`? Then you could instantiate that in your code.

spite commented:
Yes, that's just the change to be able to convert internal constants to GL constants, like in https://github.com/spite/THREE.UpdatableTexture/blob/master/UpdatableTexture.js#L51

What else could THREE.WebGLUtils provide? The whole function https://github.com/spite/three.js/blob/07239a245825a077347abef28c82dd27a81f75ea/src/renderers/WebGLRenderer.js#L2472 could be a map, and I'm not sure, but may be you don't even need to reference the actual gl context, it can be any context. That would remove the dependency to renderer.paramThreeToGL, and just use THREE.convert or THREE.WebGLUtils.convert

mrdoob commented:
Sounds good! I vote for `THREE.WebGLUtils.convert`.

spite committed:
removed

spite committed:
addd WebGLUtils

spite commented:
`WebGLUtils` added. I've tested with `UpdatableTexture` and everything works as expected.

I'm not entirely happy with [this part](https://github.com/spite/three.js/blob/615e25b578f8f3ec0fdcab0045996956143d6dba/src/extras/WebGLUtils.js#L11), because it's repeated code from [the init part of WebGLRenderer](https://github.com/spite/three.js/blob/615e25b578f8f3ec0fdcab0045996956143d6dba/src/renderers/WebGLRenderer.js#L252).

Idea: All the `if` could be replaced by a `Map` in `.convert()`.

Question: Should `WebGLUtils` replace all `threeParamToGL` calls in the code?

mrdoob commented:
> I'm not entirely happy with [this part](https://github.com/spite/three.js/blob/615e25b578f8f3ec0fdcab0045996956143d6dba/src/extras/WebGLUtils.js#L11), because it's repeated code from [the init part of WebGLRenderer](https://github.com/spite/three.js/blob/615e25b578f8f3ec0fdcab0045996956143d6dba/src/renderers/WebGLRenderer.js#L252).

Yeah, that's a bit too aggressive. I think I would do this:

```js
function WebGLUtils( gl, extensions ) {

    function convert() {}
    return { convert: convert }

}

export { WebGLUtils }
```

Then in `WebGLRenderer`:

```js
utils = new WebGLUtils( _gl, extensions );
```

```js
utils.convert( foo );
```

Then, in `UpdatableTexture` you could do this:

```js
var utils = new THREE.WebGLUtils( renderer.context, renderer.extensions );
```

> Question: Should `WebGLUtils` replace all `threeParamToGL` calls in the code?

Yep!

spite committed:
WebGLUtils uses provided WebGL context and extensions
Removed threeParamToGL and all calls

spite commented:
Much better with this design. I've updated `WebGLUtils` and replaced all calls to `threeParamToGL`. 

`WebGLTextures` and `WebGLState` had a reference to `threeParamToGL` in its constructor, I've replaced it with a reference to the renderer's `utils` object. But since they also get a reference to `gl` and `extensions`, may be they could create their own `WebGLUtils` instance. I don't think it makes a big difference one way or the other.

mrdoob commented:
Beautiful!

mrdoob commented on src/extras/WebGLUtils.js:
```diff
@@ -0,0 +1,140 @@
+ * @author thespite / http://www.twitter.com/thespite
+ */
+
+import { REVISION, MaxEquation, MinEquation, RGB_ETC1_Format, RGBA_PVRTC_2BPPV1_Format, RGBA_PVRTC_4BPPV1_Format, RGB_PVRTC_2BPPV1_Format, RGB_PVRTC_4BPPV1_Format, RGBA_S3TC_DXT5_Format, RGBA_S3TC_DXT3_Format, RGBA_S3TC_DXT1_Format, RGB_S3TC_DXT1_Format, SrcAlphaSaturateFactor, OneMinusDstColorFactor, DstColorFactor, OneMinusDstAlphaFactor, DstAlphaFactor, OneMinusSrcAlphaFactor, SrcAlphaFactor,  ...
```
Could you clean up the imports here? Also in `WebGLRenderer`?

spite committed:
moved WebGLUtils to rendereres/webgl

spite committed:
Moved WebGLUtils to renderers/webgl

spite committed:
clean up of constants

spite commented:
Moved to renderers/webgl and clean up of constants on both WebGLUtils and WebGLRenderer. 

Note: There's more constants to clean up in WebGLTextures, and probably more places

mrdoob commented:
Thanks!
